File: day7/454. 4Sum II.py
class Solution:

    def fourSumCount(self, nums1: list[int], nums2: list[int],
                     nums3: list[int], nums4: list[int]) -> int:
        hashmap = dict()
        for n1 in nums1:
            for n2 in nums2:
                if n1 + n2 in hashmap:
                    hashmap[n1 + n2] += 1
                else:
                    hashmap[n1 + n2] = 1

        count = 0
        for n3 in nums3:
            for n4 in nums4:
                if -(n3 + n4) in hashmap:
                    count += hashmap[0 - (n3 + n4)]
        return count


#  使用 get方法

    def fourSumCount(self, nums1: list[int], nums2: list[int],
                     nums3: list[int], nums4: list[int]) -> int:
        hashmap = dict()
        for n1 in nums1:
            for n2 in nums2:
                # 获取 n1+n2的值，如果不存在返回 0，
                # +1 后赋值给hashmap
                hashmap[n1 + n2] = hashmap.get(n1 + n2, 0) + 1
        count = 0
        for n3 in nums3:
            for n4 in nums4:
                key = -n3 - n4
                if key in hashmap:
                    count += hashmap[key]

        return count

File: day7/454. 4Sum II.cpp
/**
 * @Date: 2023 Nov 20
 * @Time: 01:40
 * @Author: Chris
 * @Desc: 454. 4Sum II
 * @Link: https://leetcode.cn/problems/4sum-ii/description/
 **/
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
 public:
  int fourSumCount(vector<int>& nums1, vector<int>& nums2, vector<int>& nums3,
                   vector<int>& nums4) {
    unordered_map<int, int> umap;  // key = a+b的数值, value为和出现的次数
    for (int num1 : nums1) {
      for (int num2 : nums2) {
        umap[num1 + num2]++;
      }
    }

    int count = 0;  // count a+b+c+d=0 出现的次数
    // a+b = 0-(c+d)
    for (int num3 : nums3) {
      for (int num4 : nums4) {
        // 如果不存在key,会自动插入key,并返回value=0.
        count += umap[0 - (num3 + num4)];
      }
    }
    return count;
  }
};
int main() { return 0; }